src/PdoConnection.php

```diff
@@ -248,7 +268,17 @@ public function rollBack()
      */
     public function commit()
     {
-        return $this->connect()->commit();
+        if (--$this->transactionCount === 0) {
+            if (!$this->rollbackNeeded) {
+                return $this->connect()->commit();
+            }
+
+            $this->rollbackNeeded = false;
+            $this->connect()->rollback();
```

Comment:
Yea, it feels a bit weird, but if one of the inner transactions fails, then we must not commit the outer transaction.  We return false to indicate that this has happened.